websocket/connection.go

```diff
@@ -154,50 +161,53 @@ func (c *ManagedConnection) connect() (err error) {
 	return err
 }
 
-// keepalive keeps the connection open and reads control messages.
-// All messages are discarded.
-func (c *ManagedConnection) keepalive() (err error) {
+// keepalive keeps the connection open.
+func (c *ManagedConnection) keepalive() error {
+	for {
+		if err := c.read(); err != nil {
+			return err
+		}
+	}
+}
+
+// read reads the next message from the connection.
+// If a messageChan is supplied and the current message type is not
+// a control message, the message is sent to that channel.
+func (c *ManagedConnection) read() error {
+	c.connectionLock.Lock()
+	defer c.connectionLock.Unlock()
+
+	if c.connection == nil {
+		return ErrConnectionNotEstablished
+	}
+
 	c.readerLock.Lock()
 	defer c.readerLock.Unlock()
 
-	for {
-		func() {
-			c.connectionLock.RLock()
-			defer c.connectionLock.RUnlock()
-
-			if conn := c.connection; conn != nil {
-				var reader io.Reader
-				var messageType int
-				messageType, reader, err = conn.NextReader()
-				if err != nil {
-					conn.Close()
-				}
-
-				// Send the message to the channel if its an application level message
-				// and if that channel is set.
-				if c.messageChan != nil && (messageType == websocket.TextMessage || messageType == websocket.BinaryMessage) {
-					if message, _ := ioutil.ReadAll(reader); message != nil {
-						c.messageChan <- message
-					}
-				}
-			} else {
-				err = ErrConnectionNotEstablished
-			}
-		}()
+	messageType, reader, err := c.connection.NextReader()
+	if err != nil {
+		c.connection.Close()
+		c.connection = nil
+		return err
+	}
 
-		if err != nil {
-			return err
+	// Send the message to the channel if its an application level message
+	// and if that channel is set.
+	if c.messageChan != nil && (messageType == websocket.TextMessage || messageType == websocket.BinaryMessage) {
+		if message, _ := ioutil.ReadAll(reader); message != nil {
+			c.messageChan <- message
```

Comment:
Do we care to specify the type of the message? How the client is supposed to understand whether it was binary or text message?

Reply:
I think at this point we don't care. I see the point but would rather leave a TODO for a later PR?

Reply:
SG